app/workers/messaging.php

```diff
@@ -20,59 +33,197 @@
+            default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
+        };
+
+        $recipientsId = $messageRecord->getAttribute('to');
+
+        /**
+        * @var Document[] $recipients
+        */
+        $recipients = [];
+
+        $topics = $this->dbForProject->find('topics', [Query::equal('$id', $recipientsId)]);
+        foreach ($topics as $topic) {
+            $recipients = \array_merge($recipients, $topic->getAttribute('targets'));
         }
 
-        if (empty($this->from)) {
-            Console::info('Skipped sms processing. No phone number has been set.');
-            return;
+        $users = $this->dbForProject->find('users', [Query::equal('$id', $recipientsId)]);
+        foreach ($users as $user) {
+            $recipients = \array_merge($recipients, $user->getAttribute('targets'));
         }
 
-        $message = new SMS(
-            to: [$this->args['recipient']],
-            content: $this->args['message'],
-            from: $this->from,
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
+        $recipients = \array_filter($recipients, fn (Document $recipient) => $recipient->getAttribute('providerId') === $providerRecord->getId());
+
+        $identifiers = \array_map(function (Document $recipient) {
+            return $recipient->getAttribute('identifier');
+        }, $recipients);
+
+        $maxBatchSize = $provider->getMaxMessagesPerRequest();
+        $batches = \array_chunk($identifiers, $maxBatchSize);
 
-        try {
-            $this->sms->send($message);
-        } catch (\Exception $error) {
-            throw new Exception('Error sending message: ' . $error->getMessage(), 500);
+        $results = batch(\array_map(function ($batch) use ($messageRecord, $providerRecord, $provider) {
+            return function () use ($batch, $messageRecord, $providerRecord, $provider) {
+                $deliveredTo = 0;
+                $deliveryErrors = [];
+                $messageData = clone $messageRecord;
+                $messageData->setAttribute('to', $batch);
+                $message = match ($providerRecord->getAttribute('type')) {
+                    'sms' => $this->buildSMSMessage($messageData, $providerRecord),
+                    'push' => $this->buildPushMessage($messageData),
+                    'email' => $this->buildEmailMessage($messageData, $providerRecord),
+                    default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
+                };
+                try {
+                    $provider->send($message);
+                    $deliveredTo += \count($batch);
+                } catch (\Exception $e) {
+                    foreach ($batch as $identifier) {
+                        $deliveryErrors[] = 'Failed to send message to target' . $identifier . ': ' . $e->getMessage();
+                    }
```

Comment:
I'm not sure this will be accurate, some messages in a batch might send successfully before failure

Reply:
Will address outside the scope of this PR